Update interfaces for event/configuration handling in Falaise modules

drbenmorgan opened the pull request:
This updates the following existing and new interfaces related to handling `datatools::properties` in Falaise modules:

- `falaise::property_set`
  - Convenience functions for parameter assignment
  - Add ability to validate of incoming values from scripts
- `falaise::bounded_int`
  - Very simple type to model an integral value that is required to be within a limited range of values
  - Intended for use in reconstruction for automatic and safe indexing of tracker/calorimeter blocks rather than the current bare integers with boilerplate error checking everywhere.
- `snemo::datamodels`
  - A couple of free functions to make adding data products to the event record easier and with clearer errors than provided by `datatools::things`

Whilst there's a reasonable amount of code here, I hope it's relatively clear. The primary thing to look at is the documentation in:

- source/falaise/bounded_int.h
- source/falaise/property_set.h
- source/falaise/snemo/datamodels/event.h

As these are intended for simulation/reconstruction/analysis users, this needs to be clear, so questions, feedback etc very welcome!

drbenmorgan committed:
Minimal bounded integer implementation and test

The geometry and reconstruction algorithms in Falaise have many places
where a limited range of integers is used to address "things", for
example tracker cell indices. Having values outside these ranges is
an error, and whilst these are tracked, the resulting checks are
bolierplated throughout the code.

Provide a basic "bounded_integer" type to represent a range
[lower,upper] of integral values. Constructing or assigning with a
value outside of this range results in an out of range exception.

Implement minimal unit tests of this behaviour

drbenmorgan committed:
Skeleton for new interfaces

drbenmorgan committed:
Skeleton for self-assign test/impl

drbenmorgan committed:
Free function to add data products to event with checks

An identified use case in Falaise modules is to add data products
to the event record under a given key if and only if the key does
not exist in the record. Whilst datatools::things's add member function
does check this, it happens relatively late and with a not so helpful
exception.

Provide a free template function to allow modules to add a data product
to an event record with:

- Fast runtime check on key existence with throw of dedicated existing key
  exception in case key exists.
- Compile time check that type of data product meets requirements for
  storage in the event record.

Add the compile time check to the existing getOrAddToEvent function.

drbenmorgan committed:
Compile time constraints on bounded_int

Disallow equal bounds, recommend named constant instead.
Require lower bound is strictly less than upper for consistency.

drbenmorgan committed:
Allow property_set construction from std::istream

To allow reading of data from any input stream, provide a free
function taking a istream and the property_set to be filled.

Refactor construction from file to use new istream based function.
This will allow future improvement to allow reading of property_sets
from any source (e.g. GitDB/SQLite).

drbenmorgan committed:
Add documentation and additional test

drbenmorgan committed:
Documentation fixes for paths/refs

drbenmorgan committed:
Convenience member function for self default parameters

A common pattern identified in reconstruction modules use of
property_set is "self defaulting":

```
// Usually in constructor
double a = 3.14;

a = ps.get<double>("key_for_a", a);
```

i.e. the parameter "key_for_a" is not required, and should take
'a's existing value when the key is not present.

Provide a clearer interface for this use case "assign_if", that
takes the variable to be assigned to by non-const reference.
Assign the value held at the supplied key if it exists and of
the matching type. Throw on mismatched types, leave variable
unchanged if key is missing.

drbenmorgan committed:
Provide property_set interfaces for parameter validation

Add get_if and assign_if interfaces taking a user supplied predicate
object that should take the parameter and return true/false to indicate
if it meets the required constraints.

Update tests and documentation for these new interfaces.

drbenmorgan commented:
Bumping this for review...

cherylepatrick commented:
Thanks for these @drbenmorgan - super useful! Just reading through the documentation in the header files. Got a meeting shortly so might submit these one at a time.

Bounded int:

- line 27: Really small thing: is it maybe clearer to refer to an "integer value" rather than an "integral value" in case people think it means it's something to do with integration? It might be less technically correct but I think it's also less ambiguous.
- Another small thing - line 34: tracker columns are actually [0,112] (there's a bonus middle column that connects the two halves).
- I like the example of how to declare them, that's great! And the example of declaring regular ints and then trying to use them in a bounded_int function is good too, but I think I would add some comments above line 62 to explain what this is illustrating. For example "If you pass a (regular) integer type to  a function that takes a bounded_int, it will check the boundaries for you and throw an exception if out of bounds"
- Typos on lines 36 (vunerable) and 87 (uinderflow)

cherylepatrick commented:
Property_set:
I'm a bit confused at the start of this one... sorry if I am being slow. Looking further down, I guess this is to deal with the stuff that is loaded in from the flsimulate config files etc? I think we might want a (brief) introduction to the concept of properties at the top of this file, and when you need to use them. And are they typed or not? (you replace an integer with a double at one point...)

I think a lot of what I am commenting on here is not stuff that's been changed, so feel free to ignore me :)
Line 80: I thought I didn't get this but I actually did after reading ahead a few more lines. I don't know if it's trouble, but when you generally retrieve a variable it's "get", but in this case it's "assign_if" - should it be more like "get_if"? Oh, but I see there *is* a get_if later on.  And what would happen if you had just done "get" on the "nonexistent" property? I think this also comes up later... but I'm still not entirely clear when you get vs when you assign.


Line 138: "a" contained 2 but you can't get it as a double... but 2 is a valid double, isn't it? Or at least 2.0 is? How do you find the type of one of these?

Line 149: How do these quantities with units work? Do you have to get them into 2 variables? Oh I see, you have to get them into a variable of the right type. Is there a list of the unit types? Is it always the thing it's named as with _t after it (e.g. length_t)?

Line 189/190 - it's ps.get, not table.get?

Line 197 - Seems like you shouldn't be able to use "table" as a  keyword at all if you're gonna use table.x as one? Documentation is clear to me though.

Line 218 is this the same as assign_if?

Line 230 Is there a list of standard checks or do you write your own somewhere?


Line 291 what is a property/atom?

Line 299 what does it mean by sequence?

cherylepatrick commented:
event.h: Again I think it would benefit from an example of where it gets used. This one seems pretty short but also a bit mysterious. 

- Line 21 - is it getting or setting (Adding to the event)? It's getting, right? (As there's not place to input a value)
- I'm not sure exactly what an event is here. It's not a data tools:: properties but it is still a key-value system? And it isn't an actual C++ object model where properties are variables within the object... how do we know the keys available? And their types?

Is this what something like Gamma clustering uses when it, for example, wants to add gamma particle info into the PTD bank? But then those will need to become actual objects that can be retrieved later on, so that doesn't seem right either. I'm sorry! This one is really confusing me.

cherylepatrick reviewed (changes requested):
Hi @drbenmorgan - I left some questions in the comments. I hope they are helpful! The event model one was quite confusing to me (not the changes from this PR - just the concept in general) - I'm probably being a bit dim but I suspect I'm not the only person that doesn't get it...

drbenmorgan committed:
Address review comments on documentation

drbenmorgan committed:
Address review comments on additional documentation

drbenmorgan committed:
Address review comments on code/docs

Add additional documentation in main section on an example use case
in a flreconstruct module. Clarify other sections and fix typos.

Add new member function to check type held at a key explicitly.
Implement basic test to exercise this.

drbenmorgan commented:
Thanks @cherylepatrick! I've pushed some changes to address your comments, but to answer a couple which were more general:

> Line 80: I thought I didn't get this but I actually did after reading ahead a few more lines. I don't know if it's trouble, but when you generally retrieve a variable it's "get", but in this case it's "assign_if" - should it be more like "get_if"? Oh, but I see there _is_ a get_if later on. And what would happen if you had just done "get" on the "nonexistent" property? I think this also comes up later... but I'm still not entirely clear when you get vs when you assign.

The `get` vs `assign` is to deal with the case that you have a pre-existing variable (typically a class data member) that is set to its default. When you use `get` this results in the code:

```cpp
int myVariable = 42;
...
myVariable = ps.get<int>("key", myVariable);
```

So `myVariable` is getting self-re-assigned even if "key" doesn't exist in `ps`. That's likely not a huge issue bug/performance wise, but `assign_if` allows the form:

```cpp
int myVariable = 42;
...
ps.assign_if<int>("key", myVariable);
```

Here, `myVariable` is only touched if the key exists, remove the self-re-assignment. Maybe that's not critical though. The data_model header: 

https://github.com/SuperNEMO-DBD/Falaise/blob/develop/source/falaise/snemo/datamodels/data_model.h

> Line 197 - Seems like you shouldn't be able to use "table" as a keyword at all if you're gonna use table.x as one? Documentation is clear to me though.

Bayeux's `datatools::properties` parser allows this form, and is used in some of our config files (albeit mostly for things in Bayeux's geomtools which we don't need to worry about). I've clarified the docs to indicate an exception is thrown by `property_set` if one tries to extract a "malformed table" as a "correct table".

> Is this what something like Gamma clustering uses when it, for example, wants to add gamma particle info into the PTD bank? But then those will need to become actual objects that can be retrieved later on, so that doesn't seem right either. I'm sorry! This one is really confusing me.

Exactly! The `falaise::event_record` is just a `typedef` to `datatools::things`. These functions are just wrappers around the use cases of extracting and putting data (like PTD) into the `data tools::things` instance that we pass through the pipeline as our event model. I've added docs to that effect, but not sure that will have clarified things, see what you think! 


> * I'm not sure exactly what an event is here. It's not a data tools:: properties but it is still a key-value system? And it isn't an actual C++ object model where properties are variables within the object... how do we know the keys available? And their types?

That's still a bit of an issue - like properties, there's no guarantee of bank names/types (other than key is a string, the type stored inherits from `datatools::i_serializable`.

https://github.com/SuperNEMO-DBD/Falaise/blob/develop/source/falaise/snemo/datamodels/data_model.h

Has functions to return the default bank labels for things like particle track data and so on, but that's on trust/convention rather than enforced.

drbenmorgan commented:
Ah, we do have docs on modules/config and working with data on Doxygen:

https://supernemo.org/Falaise/writingflreconstructmodules.html

https://supernemo.org/Falaise/workingwitheventrecords.html

drbenmorgan commented:
Just bumping again for review - I _think_ I addressed @cherylepatrick's comments, but see what you think!

cherylepatrick commented:
Looks good, thanks Ben
